src/providers.ts

```diff
@@ -65,6 +65,36 @@ export function CompatibleProvider(provider: any): Provider {
     });
 }
 
+/**
+ * Wrap an existing provider into one that will return a fixed getCode result for items defined in codeCache.
+ * The cache is treated as read-only, it will not be updated. Mainly used to avoid an extra RPC call when we already have the bytcode.
+ * @param provider - Any existing provider
+ * @param codeCache - Object containing address => code mappings
+ * @returns {Provider} - Provider that will return a fixed getCode result for items defined in codeCache.
+ * @example
+ * ```ts
+ * const address = "0x0000000000000000000000000000000000000001";
+ * const bytecode = "0x6001600101"
+ * const cachedProvider = WithCachedCode(provider, {
+ *   [address]: bytecode,
+ * });
+ * const code = await cachedProvider.getCode(address);
+ * console.log(code); // "0x6001600101"
+ * ```
+ */
+export function WithCachedCode(provider: Provider, codeCache: Record<string, string>): Provider {
```

Comment:
Considering making this take a getter instead of a Record, hmm.

Reply:
Maybe something like...

```ts
const cachedCodeProvider = whatsabi.providers.WithCachedCode(provider, (addr) => { if (addr === address) return bytecode; });
```

Feels more versatile, but also more footgunny. I want to avoid people just returning the same bytecode for all addresses, in case we end up having some weird proxy resolver that requires to load other addresses' bytecode and this breaks it.